vars/pushDockerImages.groovy

```diff
@@ -0,0 +1,139 @@
+  def sourceTag = version
+  def aliasVersion = ""
+  if (snapshot) {
+    // remove third number in version
+    aliasVersion = version.substring(0, version.lastIndexOf(".")) + "-SNAPSHOT"
+    sourceTag += "-SNAPSHOT"
+  }
+
+  // What docker tags are gonna be used
+  def tags = calculateTags(sourceTag, aliasVersion)
+
+  println tags
+  dockerLogin(secret: "${secret}", registry: "${registry}")
+  variants?.each { variant, sourceNamespace ->
+    tags.each { tag ->
+      // TODO:
+      // For backward compatibility let's ensure we tag only for amd64, then E2E can benefit from until
+      // they support the versioning with the architecture
+      if ("${arch}" == "amd64") {
+        doTagAndPush(registry: registry, imageName: imageName, variant: variant, sourceTag: sourceTag, targetTag: "${tag}", sourceNamespace: sourceNamespace, targetNamespace: targetNamespace)
+      }
+      doTagAndPush(registry: registry, imageName: imageName, variant: variant, sourceTag: sourceTag, targetTag: "${tag}-${arch}", sourceNamespace: sourceNamespace, targetNamespace: targetNamespace)
+    }
+  }
+}
+
+/**
+* Tag and push the source docker image. It retries to add resilience.
+*
+* @param imageName of the docker image
+* @param variant name of the variant used to build the docker image name
+* @param sourceNamespace namespace to be used as source for the docker tag command
+* @param targetNamespace namespace to be used as target for the docker tag command
+* @param sourceTag tag to be used as source for the docker tag command, usually under the 'beats' namespace
+* @param targetTag tag to be used as target for the docker tag command, usually under the 'observability-ci' namespace
+*/
+def doTagAndPush(Map args = [:]) {
+  def name = args.imageName
+  def variant = args.variant
+  def sourceTag = args.sourceTag
+  def targetTag = args.targetTag
+  def sourceNamespace = args.sourceNamespace
+  def targetNamespace = args.targetNamespace
+  def registry = args.registry
+
+  def sourceName = "${registry}/${sourceNamespace}/${name}${variant}:${sourceTag}"
+  def targetName = "${registry}/${targetNamespace}/${name}${variant}:${targetTag}"
+  def iterations = 0
+
+  waitUntil {
+    iterations++
+    def status = sh(label: "Change tag and push ${targetName}",
+                    script: """#!/bin/bash
+                      set -e
+                      echo "source: '${sourceName}' target: '${targetName}'"
+                      if docker image inspect "${sourceName}" &> /dev/null ; then
+                        docker tag "${sourceName}" "${targetName}"
+                        docker push "${targetName}"
+                      else
+                        echo "docker image ${sourceName} does not exist"
+                      fi""",
+                    returnStatus: true)
+    if (status > 0 && iterations < 3) {
+      sleep 5 * iterations
+      return false
+    } else if (status > 0) {
+      return false
+    } else {
+      return true
+    }
```

Comment:
I'm trying to follow the values in the case the SH script fails, and I see a potential infinite loop here (or at least not sure if the `waitUntil` controls it, or nothing I can check in here https://www.jenkins.io/doc/pipeline/steps/workflow-basic-steps/#waituntil-wait-for-condition)

0. SH succeeded -> `iterations=1 && status=0` -> exits at the **else** conditional (true) ✅
1. SH fails -> `iterations=1 && status > 0` -> exits at the **first** conditional (false) ✅
2. SH fails -> `iterations=2 && status > 0` -> exits at the **first** conditional (false) ✅
3. SH fails -> `iterations=3 && status > 0` -> exits at the **second** conditional (false) 🚫
4. SH fails -> `iterations=4 && status > 0` -> exits at the **second** conditional (false) 🚫
5. SH fails -> `iterations=5 && status > 0` -> exits at the **second** conditional (false) 🚫

I'd propose this eval flow (may consider printing warning log in the else clause?):
0. SH succeeded -> `iterations=1 && status=0` -> exits at the **first** conditional (true) ✅
1. SH fails -> `iterations=1 && status > 0` -> exits at the **second** conditional (false) ✅
2. SH fails -> `iterations=2 && status > 0` -> exits at the **second** conditional (false) ✅
3. SH fails -> `iterations=3 && status > 0` -> exits at the **else** conditional (true). Desired at we reached the maximum ✅
4. SH fails -> `iterations=4 && status > 0` -> Not possible, as it's covered by above case ✅
5. SH fails -> `iterations=5 && status > 0` -> Not possible, as it's covered by above case ✅
6. 
```suggestion
    if (status == 0) {
        return true
    } else if (iteration < 3) {
        return false
    } else {
        return true
    }
```

Reply:
It could be reduced to:
```groovy
if (iteration < 3) {
        return false
} else {
        return true
}
```

Reply:
BTW, removing the sleep command because IIUC `waitUntil` implies it:
> Subsequent failures will slow down the delay between attempts up to a maximum of 15 seconds.